velox/functions/lib/Re2Functions.h

```diff
@@ -255,55 +272,79 @@ template <
 struct Re2RegexpReplace {
   VELOX_DEFINE_FUNCTION_TYPES(T);
 
-  std::string processedReplacement_;
-  std::string result_;
-  std::optional<RE2> re_;
-
   FOLLY_ALWAYS_INLINE void initialize(
       const std::vector<TypePtr>& /*inputTypes*/,
       const core::QueryConfig& config,
       const arg_type<Varchar>* /*string*/,
       const arg_type<Varchar>* pattern,
       const arg_type<Varchar>* replacement) {
-    VELOX_USER_CHECK(
-        pattern != nullptr, "Pattern of regexp_replace must be constant.");
-    VELOX_USER_CHECK(
-        replacement != nullptr,
-        "Replacement sequence of regexp_replace must be constant.");
-
-    auto processedPattern = prepareRegexpPattern(*pattern);
-
-    re_.emplace(processedPattern, RE2::Quiet);
-    if (UNLIKELY(!re_->ok())) {
-      VELOX_USER_FAIL(
-          "Invalid regular expression {}: {}.", processedPattern, re_->error());
+    if (pattern != nullptr) {
+      const auto processedPattern = prepareRegexpPattern(*pattern);
+      re_.emplace(processedPattern, RE2::Quiet);
+      VELOX_USER_CHECK(
+          re_->ok(),
+          "Invalid regular expression {}: {}.",
+          processedPattern,
+          re_->error());
     }
 
-    processedReplacement_ = prepareRegexpReplacement(*re_, *replacement);
+    if (replacement != nullptr) {
+      VELOX_USER_CHECK_NOT_NULL(
+          pattern, "Pattern of regexp_replace must be constant");
```

Comment:
Does replacement being constant imply that pattern must be constant too? Could there possibly be non-constant patterns with constant replacement?

Reply:
> Does replacement being constant imply that pattern must be constant too? 

I put this restriction for simplicity.

> Could there possibly be non-constant patterns with constant replacement?

This might be possible. In that case, we would not be able to handle replacement "once" and have to process it as non-constant. So, the only change would be to replace VELOX_USER_CHECK_NOT_NULL with an 'if'. Let me know if you want me to do that.

Reply:
> In that case, we would not be able to handle replacement "once" and have to process it as non-constant. So, the only change would be to replace VELOX_USER_CHECK_NOT_NULL with an 'if'.

Yeah, this sounds good to me. Presto allows this use case, e.g., `select regexp_replace(c0, c1, '?') from (values ('12345', '[1-3]'), ('54321', '[3-5]')) t(c0, c1);`